### PR TITLE
ath79: add support for COMFAST CF-E110N V2

### PR DESCRIPTION
**Pull request by rogerpueyo**
This patch adds support for the COMFAST CF-E110N, an outdoor wireless
CPE with two Ethernet ports and a 802.11bgn radio. It will supersede https://github.com/openwrt/openwrt/pull/1453 https://github.com/openwrt/openwrt/pull/1531, which were based on the deprecated ar71xx architecture.

There are a couple of issues I haven't been able to solve:
 - The green rssimediumhigh (GPIO14) and red rssilow (GPIO11) LEDs don't seem to work, even if they did in the ar71xx target (see https://github.com/openwrt/openwrt/pull/1531). They are lit on and    off during kernel boot, but they can't be controlled afterwards.
 - The green rssihigh (GPIO16) LED is also lit on and off during kernel boot but, afterwards, it blinks to eth1's link and activity.

Any ideas? Thanks!

### CODE REVIEW

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+/ {
+	compatible = "comfast,cf-e110n", "qca,qca9533";
+	model = "COMFAST CF-E110N";
+
+	aliases {
+		serial0 = &uart;
+		led-boot = &wan;
+		led-failsafe = &wan;
+		led-upgrade = &wan;
+	};
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: You seem to have plenty of LEDs, I think, that it's better for UX if you could use different LED/colors to signal different states. You don't want user to turn off the CPE during upgrade, right?

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+	gpio_leds: leds {
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: `gpio_leds:` is probably copy&paste leftover and could be dropped.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		compatible = "gpio-leds";
+
+		wan: wan {
+			label = "cf-e110n:green:wan";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: `default-state = "off";` could be dropped

> The initial state of the LED. Valid values are "on", "off",
  and "keep". If the LED is already on or off and the default-state property is
  set the to same value, then no glitch should be produced where the LED 
  momentarily turns off (or on). The "keep" setting will keep the LED at
  whatever its current state is, without producing a glitch.  The default is
  off if this property is not present.

File: target/linux/ath79/dts/qca953x.dtsi [hunk 4]
```diff
@@ -86,7 +86,7 @@
 			pinmux: pinmux@1804002c {
 				compatible = "pinctrl-single";
 
-				reg = <0x1804002c 0x48>;
+				reg = <0x1804002c 0x44>;
```
Comment (ynezz) [inline target/linux/ath79/dts/qca953x.dtsi]: This probably needs to go into separate commit with proper explanation.

File: target/linux/ath79/image/generic.mk [hunk 5]
```diff
@@ -128,6 +128,15 @@ define Device/buffalo_wzr-hp-g450h
 endef
 TARGET_DEVICES += buffalo_wzr-hp-g450h
 
+define Device/comfast_cf-e110n
+  ATH_SOC := qca9533
+  DEVICE_TITLE := COMFAST CF-E110N
+  DEVICE_PACKAGES := rssileds kmod-leds-gpio -swconfig -uboot-envtools
+  IMAGE_SIZE := 16192k
+  SUPPORTED_DEVICES += cf-e110n
```
Comment (ynezz) [inline target/linux/ath79/image/generic.mk]: You can drop `SUPPORTED_DEVICES` as `SUPPORTED_DEVICES := comfast,cf-e110n` it's added automatically in `Device/Default` in target/linux/ath79/image/Makefile.

File: target/linux/ath79/image/generic.mk [hunk 6]
```diff
@@ -128,6 +128,15 @@ define Device/buffalo_wzr-hp-g450h
 endef
 TARGET_DEVICES += buffalo_wzr-hp-g450h
 
+define Device/comfast_cf-e110n
+  ATH_SOC := qca9533
+  DEVICE_TITLE := COMFAST CF-E110N
+  DEVICE_PACKAGES := rssileds kmod-leds-gpio -swconfig -uboot-envtools
```
Comment (ynezz) [inline target/linux/ath79/image/generic.mk]: In the description you're saying:

> The CF-E110N is very similar to the CF-E320N V2

Looking at `cf-e320n-v2` in `ar71xx` I can see `DEVICE_PACKAGES += kmod-usb-core kmod-usb2` is there any reason why you don't include them also?

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts (hunk 1 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: Good idea. Let's use the red rrsilow LED:

```
	aliases {
		serial0 = &uart;
		led-boot = &wan;
		led-failsafe = &wan;
		led-upgrade = &rssilow;
	};
```

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts (hunk 2 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: Correct.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts (hunk 3 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: Great. Then, let's drop them all. In any case, they should all later be controlled by the LED subsystem.

File: target/linux/ath79/dts/qca953x.dtsi (hunk 4 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca953x.dtsi]: Sorry, it didn't mean to go in the commit and was later reverted in a following push. I was trying different values, out of desperation.

File: target/linux/ath79/image/generic.mk (hunk 6 above)
Comment (rogerpueyo) [reply target/linux/ath79/image/generic.mk]: The CF-E320N-V2 is based on the QCA9531 SoC and the CF-E110N is based on the QCA9533, which I read (https://forum.archive.openwrt.org/viewtopic.php?id=69479) it has no USB. Therefore I thought it was not needed.

Comment (ynezz) [reply target/linux/ath79/image/generic.mk]: Ok, I'm not sure, but I think, that I've seen some USB nodes in this DTS, that's why I was confused.

File: target/linux/ath79/image/generic.mk (hunk 5 above)
Comment (rogerpueyo) [reply target/linux/ath79/image/generic.mk]: Done.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	};
+
+	ath9k-leds {
+		compatible = "gpio-leds";
+
+		wlan {
+			label = "cf-e110n:green:wlan";
+			gpios = <&gpio 0 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+			default-state = "off";
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: This one could go as well.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		led-failsafe = &wan;
+		led-upgrade = &rssilow;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wan: wan {
+			label = "cf-e110n:green:wan";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: You're using GPIOs 0,2,3 in your DTS, in your ar71xx implementation you've had `ath79_gpio_function_enable(AR934X_GPIO_FUNC_JTAG_DISABLE);` but I don't see the `jtag_disable_pins` pinctrl being used anywhere.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+		};
+
+		rssimediumhigh {
+			label = "cf-e110n:green:rssimediumhigh";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		rssihigh {
+			label = "cf-e110n:green:rssihigh";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: > The green rssihigh (GPIO16) LED is also lit on and off during kernel

I've some older version of DTS downloaded where you had `gpio 17`, so it's fixed now?

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+			linux,default-trigger = "phy0tpt";
+			default-state = "off";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+
+		poll-interval = <20>;
+		pinctrl-names = "default";
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: `pinctrl-names = "default";` is probably another left-over, right?

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts (hunk 8 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: You're right. I had it in some previous version but then I started to add, remove and change stuff to see if I could catch what was going wrong with the GPIO pins, so I must have removed it.

```
keys {
		compatible = "gpio-keys-polled";
		poll-interval = <20>;

		pinctrl-0 = <&jtag_disable_pins>;
		
		reset {
			label = "Reset button";
			linux,code = <KEY_RESTART>;
			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
			debounce-interval = <60>;
		};
	};
```

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts (hunk 10 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: You're completely right.

**Comment by rogerpueyo** [comment]
@ynezz Thank you very much, again, for your review.

I have detected the following behavior with GPIOs 14 and 11. If I don't add them as LEDs in the DTS, i can enable them with:

```
cd /sys/class/gpio
echo 11 > export
echo 14 > export
```

then, if I set their direction to "out", no matter their value, they don't seem to work. However, if I set them to "in", they enter "floating" state in which the corresponding LEDs dimly glow. So, the LED numbers are correct (11 and 14), but I imagine that something is disabling/overwriting them when in "out" direction.

And then there is GPIO 16, which blinks according to eth1 link and activity, no matter how I configure it. By the way, I have observed that the same behavior occurs in U-boot: rrsihigh (GPIO16) blinks to eth1 link and activity. Could there be something wrong in this specific SoC's support when configuring the GPIOs?

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts (hunk 9 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: Yes, my bad. GPIO 16 is correct; GPIO 17 corresponds to the reset button.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+/ {
+	compatible = "comfast,cf-e110n", "qca,qca9533";
+	model = "COMFAST CF-E110N";
+
+	aliases {
+		serial0 = &uart;
+		led-boot = &wan;
+		led-failsafe = &wan;
+		led-upgrade = &rssilow;
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: @rogerpueyo I'm sorry, that I've told you about this LED change. I did the same and [it was rejected](http://lists.infradead.org/pipermail/openwrt-devel/2018-December/015119.html), and the reason is, that it's not good idea to use RSSI LEDs for system states.

**Comment by ynezz** [comment]
> Could there be something wrong in this specific SoC's support when configuring the GPIOs?

Yes, everything is possible, but I quite don't get it, because you're saying that it was working for you in ar71xx, and in ath79 it's not. There is `qca9533_tplink_tl-wr841-v11.dts` already, which seems to use `gpio 1,2,3` without `jtag_disable_pins` pinctrl and using `gpio 11,14,16` as well for LEDs. Looking at the ar71xx code for that tplink I don't see `ath79_gpio_function_enable(AR934X_GPIO_FUNC_JTAG_DISABLE);` either. I'm confused as well :-)

Maybe there are bootloaders or something else setting some registers in gpio/pinmux in different ways and you probably need to dump and diff the registers in ar71xx (working case) and in ath79 (non working case) and find out where the problem is. I can't help you more with this as I don't own such device, sorry.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts (hunk 11 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/qca9533_comfast_cf-e110n.dts]: No worries :)

**Comment by rogerpueyo** [comment]
Regarding the `jtag_disable_pins` setting, GPIOs 0-3 work no matter I add it or not (ath79). So, I guess, JTAG in those pins is forcibly disabled somewhere else.
I have tried the TL-WR841N v11 initramfs on my CF-E110N and the LEDs on GPIOs 0-3 work, those on GPIOs 11, 14 and 16 don't. Actually, _I doubt_ they work in the actual device (I don't have any around to test).

Definitely, the bootloader **is** doing something with the registers (in u-boot, LED on GPIO16 blinks to eth1 link). I assume that in ar71xx the registers are properly reset/reconfigured/whatever somewhere, while in ath79 they are not, but fixing this goes beyond my knowledge. :(

> I can't help you more with this as I don't own such device, sorry.
Indeed, I appreciate your comments and help so far very much :)

**Comment by rogerpueyo** [comment]
I've been reading the QCA9531's datasheet (please find it [here](https://github.com/Deoptim/atheros/)). There the following registers are described in pages 110 and 111:
- 7.4.14 GPIO Function 2 (GPIO_OUT_FUNCTION2)
- 7.4.15 GPIO Function 3 (GPIO_OUT_FUNCTION3)
- 7.4.16 GPIO Function 4 (GPIO_OUT_FUNCTION4)
which refer to GPIOs 11, 12&14 and 16, respectively.

In u-boot, I perform the following memory register dumps:
```
ath> md 0x18040034 1; md 0x18040038 1; md 0x1804003c 1
18040034: 2c160000    ,...
18040038: 2a2b0000    *+..
1804003c: 00000129    ...)
```

Now, I write zeros to the registers corresponding to GPIOs 11, 12, 14 and 16, and dump them again:
```
ath> mw 0x18040034 00160000; mw 0x18040038 2a000000; mw 0x1804003c 00000100
ath> md 0x18040034 1; md 0x18040038 1; md 0x1804003c 1
18040034: 00160000    ....
18040038: 2a000000    *...
1804003c: 00000100    ....
```

Well, if I then load the kernel and boot to OpenWrt (without rebooting the device), I get the GPIOs to work as LEDs, the same way as they do in the ar71xx image. If I reboot, U-boot resets the registers to the previous values and the LEDs do not work as they should.

So... problem not solved but, at least, very narrowed! :)

@ynezz Where/Who do you think I could ask for help regarding this issue? Mailing list? Forums? You seem to have some experience :)

**Comment by ynezz** [comment]
@rogerpueyo 

> 0x0 is the value I want to write

That's wrong, you need to set the bits to 1, don't you?

```
&pinmux {
        led_rssilow_pin: pinmux_rssilow_pin {
                pinctrl-single,bits = <0x8 0xff000000 0xff000000>;
        };

        led_rssimediumhigh_pin: pinmux_rssimediumhigh_pin {
                pinctrl-single,bits = <0xc 0xff0000 0xff0000>;
        };

        led_rssihigh_pin: pinmux_rssihigh_pin {
                pinctrl-single,bits = <0x12 0xff 0xff>;
        };
};
```

> Then, looking at the jtag_disable_pins entry in qca953x.dts I saw it is quite different, from my 
attemps

Because it's setting DISABLE_JTAG (bit 1) in 0x1804006C to 1 (so the value and mask is 2).

**Comment by ynezz** [comment]
@rogerpueyo 

I've just noticed this from your example:

```
leds {
    pinctrl-1
```

and this is probably not going to work as you need to use `pinctrl-0` first (required property). Could you please check under /sys that the pinctrl nodes are actually being used? Maybe there's some clue in the dmesg?

**Comment by rogerpueyo** [comment]
You're right, it says "pinctrl-1", but I've also tried "pinctrl-0", putting everything together with the &jtag, moving it up and down... If I use "pinctrl-0":
 - In /sys:
```
root@OpenWrt:~# ls -la /sys/firmware/devicetree/base/ahb/apb/pinmux@1804002c
-r--r--r--    1 root     root             4 Dec 16 12:58 #pinctrl-cells
-r--r--r--    1 root     root             4 Dec 16 12:58 #size-cells
drwxr-xr-x    6 root     root             0 Dec 16 12:58 .
drwxr-xr-x   14 root     root             0 Dec 16 12:58 ..
-r--r--r--    1 root     root            15 Dec 16 12:58 compatible
-r--r--r--    1 root     root             7 Dec 16 12:58 name
-r--r--r--    1 root     root             0 Dec 16 12:58 pinctrl-single,bit-per-mux
-r--r--r--    1 root     root             4 Dec 16 12:58 pinctrl-single,function-mask
-r--r--r--    1 root     root             4 Dec 16 12:58 pinctrl-single,register-width
drwxr-xr-x    2 root     root             0 Dec 16 12:58 pinmux_jtag_disable_pins
drwxr-xr-x    2 root     root             0 Dec 16 12:58 pinmux_rssihigh_pin
drwxr-xr-x    2 root     root             0 Dec 16 12:58 pinmux_rssilow_pin
drwxr-xr-x    2 root     root             0 Dec 16 12:58 pinmux_rssimediumhigh_pin
-r--r--r--    1 root     root             8 Dec 16 12:58 reg
```

 - in dmesg:
```
[    0.494515] pinctrl-single 1804002c.pinmux: 576 pins at pa b804002c size 72
```

and that's it. :(

I've also tried creating two different, non-overlapping pinmux and pinmux2 sections, one for the previous jtag_disable part, another for tweaking the GPIOs, but that didn't help either. Directly on the SoC's .dtsi file:

```
			pinmux: pinmux@1804006c {
				compatible = "pinctrl-single";

				reg = <0x1804006c 0x8>;

				#size-cells = <0>;

				pinctrl-single,bit-per-mux;
				pinctrl-single,register-width = <32>;
				pinctrl-single,function-mask = <0x1>;
				#pinctrl-cells = <2>;

				jtag_disable_pins: pinmux_jtag_disable_pins {
					pinctrl-single,bits = <0x00 0x2 0x2>;
				};
			};

			pinmux2: pinmux@1804002c {
				compatible = "pinctrl-single";

				reg = <0x1804002c 0x40>;

				#size-cells = <0>;

				pinctrl-single,bit-per-mux;
				pinctrl-single,register-width = <32>;
				pinctrl-single,function-mask = <0x0>;
				#pinctrl-cells = <2>;
			};

```

And use different `pinctrl-single,function-mask` masks, etc. I couldn't make it either. :(

**Comment by rogerpueyo** [comment]
Wrap-up:

- Renamed to CF-E110N v2. The manufacturer advertises, ships and labels the device as simply "CF-E110N", but the bootloader's bootargs and the stock firmware's "board_name" identify it as "v2", so let's stick to it.
- All LEDs now working

So, ready to be merged, I hope.

**Comment by rogerpueyo** [comment]
Wrap-up:

- Renamed to CF-E110N v2. The manufacturer advertises, ships and labels the device as simply "CF-E110N", but the bootloader's bootargs and the stock firmware's "board_name" identify it as "v2", so let's stick to it.
- All LEDs now working

So, ready to be merged, I hope.

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n-v2.dts [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+			label = "Reset button";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+};
+
+&pinmux {
+	led_rssilow_pin: pinmux_rssilow_pin {
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n-v2.dts]: @rogerpueyo Whitespace issues

File: target/linux/ath79/dts/qca9533_comfast_cf-e110n-v2.dts [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+		pinctrl-single,bits = <0x8 0x0 0xff000000>;
```
Comment (ynezz) [inline target/linux/ath79/dts/qca9533_comfast_cf-e110n-v2.dts]: @rogerpueyo Whitespace issues

**Comment by mkresin** [comment]
merged, thanks!
